solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -0,0 +1,155 @@
+"""
+Provides preprocessing steps to be performed on the timeseries data.
+"""
+
+import pandas as pd
+
+from solarforecastarbiter import datamodel
+from solarforecastarbiter.validation import quality_mapping
+
+
+def apply_validation(data, qfilter, handle_func):
+    """
+    Apply validation steps based on provided filters to the data.
+
+    Parameters
+    ----------
+    qfilter : list of
+        :class:`solarforecastarbiter.datamodel.QualityFlagFilter`
+    data : :class:`pd.DataFrame`
+        Pandas DataFrame of observations and forecasts.
+    handle_func : function
+        Function that handles how `quality_flags` will be used.
+
+    Returns
+    -------
+    pd.DataFrame
+    """
+    validated_data = {}
+
+    # List of flags from filter
+    if not isinstance(qfilter, datamodel.QualityFlagFilter):
+        return TypeError(f"{qfilter} not a QualityFlagFilter")
+    filters = qfilter.quality_flags
+
+    # Apply handling function to quality flags
+    for model, values in data.items():
+
+        # Skip if empty
+        if isinstance(values, pd.DataFrame) and values.empty:
+            validated_data[model] = values.value
+            continue
+
+        # Apply only to Observations
+        if isinstance(model, datamodel.Observation):
+            validation_df = quality_mapping.convert_mask_into_dataframe(
+                values['quality_flag'])
+            validation_df = validation_df[filters]
+            validated_data[model] = handle_func(values.value,
```

Comment:
I can change the name. Just the idea is that any fillin/exclude function can be applied simply. I suppose we could use a string argument as well that maps to the function if you prefer.

Sure I will change it so that the full DataFrame is passed.